staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/list_work_estimator.go

```diff
@@ -82,23 +82,22 @@ func (e *listWorkEstimator) estimate(r *http.Request) WorkEstimate {
 		return WorkEstimate{Seats: maximumSeats}
 	}
 
-	// TODO: For resources that implement indexes at the watchcache level,
-	//  we need to adjust the cost accordingly
+	limit := numStored
+	if utilfeature.DefaultFeatureGate.Enabled(features.APIListChunking) && listOptions.Limit > 0 &&
+		listOptions.Limit < numStored {
+		limit = listOptions.Limit
```

Comment:
This isn't true for listing from cache - see my comment in the doc.

Reply:
Oh wow, I hadn't noticed.  That seems like pretty strange behavior to me.  Is it a bug?  Given that we are struggling with the costs of long list results, maybe the cacher should limit its response in this case?

Reply:
Revised estimator to match current code.

Reply:
> Is it a bug? Given that we are struggling with the costs of long list results, maybe the cacher should limit its response in this case?

It was intentional (although I agree it isn't intuitive).
The reasoning behind it is that in large clusters, when node agents were listing stuff from etcd (say e.g. to restore after control-plane outage when those lists where somewhat coordinated across many nodes), it was completely killing etcd.
And given listing from watchcache is explicit opt-in, then ignoring limit is slightly less risky, as people need to explicitly opt-in for that behavior so hopefully they understand the consequences.

With P&F giving us a better protection (once we have a good list support) [and couple other things that happened over last years] we may be able to revise it, but that requires much deeper testing and should be its own effort.

[FTR - once solve, we would be able to actually graduate https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/365-paginated-lists to stable, as solving somehow the pagination aspect for watchcache is currently the only blocker IIRC.]